edgelet/hsm-sys/azure-iot-hsm-c/src/certificate_info.c

```diff
@@ -27,170 +32,105 @@ typedef struct CERT_DATA_INFO_TAG
+static const char END_CERT_MARKER[]   = "-----END CERTIFICATE-----";
+static size_t END_CERT_MARKER_LEN     = sizeof(END_CERT_MARKER) - 1;
+
+// openssl ASN1 time format defines
+#define ASN1_TIME_STRING_UTC_FORMAT 0x17
+#define ASN1_TIME_STRING_UTC_LEN 13
+
+static void extract_first_cert_and_chain
+(
+    const char* pem_chain,
+    CERT_MARKER* first_cert,
+    CERT_MARKER* chain_cert
+)
 {
-    BUFFER_HANDLE result;
-    const char* iterator = cert_info->certificate_pem;
-    char* cert_base64;
-    size_t cert_idx = 0;
-
-    // Allocate enough space for the certificate,
-    // no need to do append a +1 due to we're not
-    // copying the headers
-    size_t len = strlen(iterator);
-    if ((cert_base64 = (char*)malloc(len)) == NULL)
-    {
-        LogError("Failure allocating base64 decoding certificate");
-        result = NULL;
-    }
-    else
+    const char *iterator = pem_chain;
+    size_t pem_len = strlen(pem_chain);
+    const char* pem_end = pem_chain + pem_len - 1;
+    const char* first_start = NULL;
+    const char* first_end = NULL;
+    const char* chain_start = NULL;
+    const char* chain_end = NULL;
+    bool first_cert_begin_marker_found = false;
+    bool first_cert_found = false;
+
+    while (*iterator != '\0')
     {
-        bool begin_hdr_end = false;
-        int begin_hdr_len = 0;
-        memset(cert_base64, 0, len);
-        cert_info->first_cert_start = cert_info->certificate_pem;
-        // If the cert does not begin with a '-' then
-        // the certificate doesn't have a header
-        if (*iterator != '-')
-        {
-            begin_hdr_end = true;
-        }
-        while (*iterator != '\0')
+        const char *ptr;
+
+        if (!first_cert_found)
         {
-            if (begin_hdr_end)
+            // check for first cert
+            if (!first_cert_begin_marker_found)
             {
-                // Once we are in the header then, copy the cert excluding \r\n on all lines
-                if (*iterator != '\r' && *iterator != '\n')
+                if ((ptr = strstr(iterator, BEGIN_CERT_MARKER)) != NULL)
                 {
-                    cert_base64[cert_idx++] = *iterator;
+                    first_start = ptr;
+                    first_cert_begin_marker_found = true;
+                    iterator = ptr + BEGIN_CERT_MARKER_LEN - 1;
                 }
-                if (*iterator == '\n' && *(iterator + 1) == '-')
+            }
+            else
+            {
+                if ((ptr = strstr(iterator, END_CERT_MARKER)) != NULL)
                 {
-                    // mark the end of the first certificate including \r\n characters
-                    cert_info->first_cert_end = iterator + END_HEADER_LENGTH + 1;
-                    if (*(cert_info->first_cert_end) == '\r')
+                    first_end = ptr + END_CERT_MARKER_LEN - 1;
+                    if ((first_end + 1 <= pem_end) && (*(first_end + 1) == '\r'))
```

Comment:
Do we need to worry about a string with `-----END CERTIFICATE-----\n\r` (line feed and carriage return are in reverse order)?

Reply:
I don't think so. If that were to happen, the first char in the chain would be \r which technically is ignored by certificate parsers.